set-balance.md

```diff
@@ -168,6 +192,346 @@ A `Result` is considered an `Action`.
+
+The dispatch origin for this call must be `Signed` by the transactor.
+
+```k
+    syntax ExitenceRequirement ::= ".AllowDeath"
+                                 | ".KeepAlive"
+
+    syntax Action ::= transfer(AccountId, AccountId, Int)
+ // ---------------------------------------------------------------------
+    rule [transfer-self]:
+         <k> transfer(ORIGIN, ORIGIN, _) => . ... </k>
+
+    rule [transfer-existing-account]:
+         <k> transfer(ORIGIN, DESTINATION, AMOUNT)
+          => set_free_balance(ORIGIN, SOURCE_BALANCE -Int AMOUNT -Int FEE)
+          ~> set_free_balance(DESTINATION, DESTINATION_BALANCE +Int AMOUNT)
+         ...
+         </k>
+         <totalIssuance> ISSUANCE => ISSUANCE -Int FEE </totalIssuance>
+         <transferFee> FEE </transferFee>
+         <accounts>
+           <account>
+             <accountID> ORIGIN </accountID>
+             <freeBalance> SOURCE_BALANCE </freeBalance>
+             ...
+           </account>
+           <account>
+             <accountID> DESTINATION </accountID>
+             <freeBalance> DESTINATION_BALANCE </freeBalance>
+             ...
+           </account>
+         </accounts>
+      requires ORIGIN =/=K DESTINATION
+       andBool DESTINATION_BALANCE >Int 0
+       andBool SOURCE_BALANCE >=Int (AMOUNT +Int FEE)
+       andBool ensure_can_withdraw(ORIGIN, Transfer, SOURCE_BALANCE -Int AMOUNT -Int FEE)
+
+    rule [transfer-create-account]:
+         <k> transfer(ORIGIN, DESTINATION, AMOUNT)
+          => set_free_balance(ORIGIN, SOURCE_BALANCE -Int AMOUNT -Int CREATION_FEE)
+          ~> set_free_balance(DESTINATION, AMOUNT)
+         ...
+         </k>
+         <totalIssuance> ISSUANCE => ISSUANCE -Int CREATION_FEE </totalIssuance>
+         <existentialDeposit> EXISTENTIAL_DEPOSIT </existentialDeposit>
+         <creationFee> CREATION_FEE </creationFee>
+         <accounts>
+           <account>
+             <accountID> ORIGIN </accountID>
+             <freeBalance> SOURCE_BALANCE </freeBalance>
+             ...
+           </account>
+           <account>
+             <accountID> DESTINATION </accountID>
+             <freeBalance> 0 </freeBalance>
+             <reservedBalance> 0 </reservedBalance>
+             ...
+           </account>
+         </accounts>
+      requires ORIGIN =/=K DESTINATION
+       andBool SOURCE_BALANCE >=Int (AMOUNT +Int CREATION_FEE)
+       andBool EXISTENTIAL_DEPOSIT >=Int AMOUNT
+       andBool ensure_can_withdraw(ORIGIN, Transfer, SOURCE_BALANCE -Int AMOUNT -Int CREATION_FEE)
+```
+
+Force a transfer from any account to any other account.  This can only be done by root.
+
+```k
+    syntax Action ::= "force_transfer" "(" Origin "," AccountId "," AccountId "," Int ")"
+ // ----------------------------------------------------------------------------------------
+    rule [force-transfer]:
+         <k> force_transfer(.Root, SOURCE, DESTINATION, AMOUNT) => transfer(SOURCE, DESTINATION, AMOUNT) ... </k>
```

Comment:
How does this end up differing from just doing a normal `transfer`?

Also, is there only a single root account, so you're choosing to call it `.Root`? Or should there be many root accounts, in which case we should check containment in a `<root-accounts>` cell?

Reply:
There is only one root (superuser) account, so `.Root` is correct.

The reason for the difference is that lower-level functions take an `AccountId`, while higher-level ones take an `Origin`.  An `Origin` can be either an `AccountId` (for a real account), `.Root` (transactions initiated by the runtime itself), or `.None` (anonymous transactions).